Refactor of Remote Flow section (#630)
italia/eidas-it-wallet-docs: docs/en/remote-flow.rst

**Steps 1-2**: The User requests to access to a protected resource of the Relying Party.

**Steps 3-5**: The Relying Party creates a state value bound to the user-agent (e.g., using an HTTP secured cookie), the Request Object available for download at the ``request_uri`` location. It then inspects the user-agent to determine whether the flow occurs on the same device as the user-agent.

**Steps 6-9 (Authorization Request)**: The Relying Party provides the user-agent with a JavaScript page inspecting the status endpoint and the Wallet Instance with a URL containing the Authorization Request. 

  In the **Cross Device Flow**, the Request URI is presented as a QR Code displayed to the User. The User scans the QR Code using the Wallet Instance and retrieves a URL with the parameters ``client_id``, ``request_uri``, ``state`` and ``request_uri_method``. 

  Below is represented a non-normative example of a QR Code issued by the Relying Party.

  .. figure:: ../../images/verifier_qr_code.svg
      :figwidth: 50%
      :align: center

  Below is represented a non-normative example of the QR Code raw payload:

  .. code-block:: text

>     https://wallet-solution.digital-strategy.europa.eu/authorization?client_id=https%3A%2F%2Frelying-party.example.org&request_uri=https%3A%2F%2Frelying-party.example.org&request_uri_method=post

Reviewer: ```suggestion
    https://wallet-solution.example.org/authorization?client_id=https%3A%2F%2Frelying-party.example.org&request_uri=https%3A%2F%2Frelying-party.example.org&request_uri_method=post
```

Remote Flow
===========

Depending on whether the User is using a mobile device or a workstation, the Relying Party MUST support the following remote flows:

* **Same Device**, the Relying Party MUST provide an HTTP location to the Wallet Instance using a redirect (302) or an HTML href in a web page;
* **Cross Device**, the Relying Party MUST provide a QR Code which the User frames with the Wallet Instance.

Once the Wallet Instance establishes the trust with the Relying Party and evaluates the request, the User gives the consent for the disclosure of the Digital Credentials, in the form of a Verifiable Presentation.

> .. _fig_High-Level-Flow-Presentation:

Reviewer 2: suggestion: move step 9 "Local User authentication" before step 8 "Scan QR Code".

link below: https://www.plantuml.com/plantuml/uml/fLPDRnit4BtpLmpKGsqWTfCU1RY8exgsaxHMBIK-r8L0xOuarYjoImvNKOlyzmx95SEr7I103L-Md7dp-EPZzDgHjEItZP2UZFRR9LeXn7b5niBeWqCx0kd02x4qdzB1zmv-navLw4RmtNkrtj2RngDe6KTLgfMQO7GdcmO9hZN3wmeZncveuXdgHAvHKueW1UeQE4eVwXQRlT9hcBBJ_gkHXcRt-kZctDC69erYuEE8t3toO4ha7N2qrgW-kiCSdVzaHQgJXE25XvEVcDiOxd09CxJXtQ1bfzQIbD65gHGrbeuoU7OuXblyswCZc5f3M17MVEACjnKASNR6n-YOQuI3wKOwXBEpZ1lOTtY0cbWCPn8SnJl9SjphfV7CdGnyaHGgOnxKCwusaLbugy0c6gKL9wnbZYKIxLkEmUWr1lfGMUCSVqGbZ0KdjnW_Cr2GQqFR1bf6W9tYVFn-Ep5rDDzON34xVh3IQWsqGV2sYLRtkeffMwZwHKdqHNhk2_YZE5niaJQc_hDKdRKIlV1-3BEDsKNy53onOGOSf7g85v7BqXCPFKYU527lZB8RmtjJoGP2tOFqxAH9LOzIlnDYROrlMMFZ3iq1HZTxiOE4CgkaPXx0H6AbYNnp4P8OKCXwHr67o2rB8OHi2DGAdZuULAZStcbwCviB083T8PbSS9sg4VfcP1zPIv909aOey78pH812p_YTowCWkycmX9LMGQDAjdAf6aKAdS36uJUGlhoOXvHeGz2cFX3n1SFMYU6SMcki3u1Wbd-5QbnPi_twQwBI7VQvV1xYSG4zpSnWnC2l1ncbhg4rN4mANNRwY5QjzZ6I96yHpAfWdnzqaq9dBzcekcXFjlmuFNB7oS1_4M9bR6YUADHmIYMWYT4hPRTO3uTH5ZGK2OG9QQpw9vOnGxis8CAzJZdfsWMP1zIvxTfYGYsSNswLSqK3HDcUhv4aXvBB1eTpBWOedB80HM6T4w5BzgTCBdsM1drzMoQk78GPVr3x5-JZwCbtdAViLUmazebrPx5ZClLfmQOFDR4Ceqd91gwTysXX7dgqXyoKtzxDONR8Tu72z2fqBdLpXXB2SbG4qw5bCzvhrvTNyylRYulVnFCUOm4dCDaW3wqhqzJizWxty4RfceioKc4_VZZ_Jurp69C3teLR400HhlLBhbHuOC-qBfIEvyKRoK2NgZTcRfgIT2f2-8jFP6MLrA-LgnhZi0P9PDNI4xf8eueSkUbPwDoJYWD0wXUB83kfcb0Efw4tenAYaEn3cuiWJjp3OuxIEWCVx-jZSPYuCxd5my2DnSlHE68Lvq3H-LVp-VI7LwUluCUNB-7c7QVmtr5fEWeRYrgNTn7OlYqMbQaHFlUTA9cZYmI7aShTmGzxVY7O9tIDfszaNFRpRZYo-etdM3zuomBFAgjQIjaOFrfQiQLZgn-507aKnYSdaOz5ndOeN2BavV_MyBDWsYLq4chmGoc17HPKctymvLygmv6JgKsVMZvQw4F5NdDelst-0m00

.. figure:: ../../images/High-Level-Flow-Presentation.svg
    :figwidth: 100%
    :align: center
    :target: https://www.plantuml.com/plantuml/svg/TL9TJy8m57tlhpZXHLcYYz-61uCXnF34d11UJCZS6bQPRMtlRF3NspB0JUBJidlEFH-v7LhA3DKV5TF-AtAXCqdeBRAgueI9zB3CUG-PXUjIKbvjX5mXySFDbc0qOqRZx05kW8jpHD5ZJQKouZiZeIHI_bbpIr64KmVJ_2nh8_gWqgXwLVfX8GpF2ShWEKMk2WwRPnAlafHdSSHn4-t4eYi-beLMGb8SC-P21gC7k0mXThQOfvDsXAVnBDWaqvTP7mVrDF7AxOssxg7SrR6krKfQtdJR8zEtTr-cpvZRxLs7lSK4evBdQ-l9lz1DWEQM6uo2a2IFjfhS1ZXa_LExQ_obbwJMN1uNQbZ_D0e6TzjAIJlQeUvzm3htxlWg7QBuispWzYTi3ilOaCl2lwuV
    
    High Level Remote Protocol Flow

A High-Level description of the remote flow, from the User's perspective, is given below and shown in :ref:`fig_High-Level-Flow-Presentation`:

  1. *Authorization Request*: the Wallet Instance obtains a URL in the Same Device flow or a QR Code containing the URL in Cross Device flow where the signed presentation Request Object is available for download.

>   2. *Redirect URI Request*: the Wallet Instance extracts from the payload the following parameters: ``client_id``, ``request_uri``, ``state``, ``request_uri_method``.

Reviewer 2: ```suggestion
  2. *Request URI Request*: the Wallet Instance extracts from the payload the following parameters: ``client_id``, ``request_uri``, ``state``, ``request_uri_method``.
```

    * If ``request_uri_method`` is provided and set with the value ``post``, the Wallet Instance SHOULD transmit its metadata to the Relying Party's ``request_uri`` endpoint using the HTTP POST method.
    * If ``request_uri_method`` is set with the value ``get`` or not present, the Wallet Instance MUST fetch the signed Request Object using an HTTP request with method GET to the endpoint provided in the ``request_uri`` parameter.

>   3. *Redirect URI Response*: the Relying Party returns a signed Request Object to the Wallet Instance. 

Reviewer 2: ```suggestion
  3. *Request URI Response*: the Relying Party returns a signed Request Object to the Wallet Instance. 
```

    
    Remote Protocol Flow

The details of each step shown in the previous picture are described below.

**Steps 1-2**: The User requests to access to a protected resource of the Relying Party.

> **Steps 3-5**: The Relying Party creates a state value bound to the user-agent (e.g., using an HTTP secured cookie), the Request Object available for download at the ``request_uri`` location and inspects the user-agent to determine whether the flow occurs on the same device of the user-agent or not.

Reviewer 2: ```suggestion
**Steps 3-5**: The Relying Party creates a state value bound to the user-agent (e.g., using an HTTP secured cookie), the Request Object available for download at the ``request_uri`` location. It then inspects the user-agent to determine whether the flow occurs on the same device as the user-agent.
```

      :figwidth: 50%
      :align: center

  Below is represented a non-normative example of the QR Code raw payload:

  .. code-block:: text

    https://wallet-solution.digital-strategy.europa.eu/authorization?client_id=https%3A%2F%2Frelying-party.example.org&request_uri=https%3A%2F%2Frelying-party.example.org&request_uri_method=post

  .. note::
      The *error correction level* chosen for the QR Code MUST be Q (Quartily - up to 25%), since it offers a good balance between error correction capability and data density/space. This level of quality and error correction allows the QR Code to remain readable even if it is damaged or partially obscured.

  Conversely, in the **Same Device Flow**, the Relying Party uses an HTTP response redirect (with status code set to 302) or an html page with an href button, containing the URL providing the same information as in the Cross-Device Flow. Below is a non-normative example:

  .. code-block:: http 

      HTTP/1.1 302 Found
      Location: https://wallet-solution.digital-strategy.europa.eu?client_id=https%3A%2F%2Frelying-party.example.org%2Fcb&request_uri=https%3A%2F%2Frelying-party.example.org%2Frequest_uri&request_uri_method=post

> HTTP/1.1 201 Created

Reviewer 2: Are lines 88 -- 91 misplaced?

      "dcql_query": {
        "credentials": [
          {
            "id": "personal id data",
            "format": "dc+sd-jwt",
            "meta": {
              "vct_values": [ "https://pidprovider.example.org/v1.0/personidentificationdata" ]
            },
            "claims": [
                {"path": ["given_name"]},
                {"path": ["family_name"]},
                {"path": ["personal_administrative_number"]}
            ]
          },
          {
            "id": "wallet attestation",
            "format": "dc+sd-jwt",
            "meta": {
              "vct_values": ["https://itwallet.registry.example.it/WalletAttestation"]
            },
            "claims": [
                {"path": ["wallet_link"]},
                {"path": ["wallet_name"]}
            ]
          }
        ]
      },
      "response_uri": "https://relying-party.example.org/response_uri",
      "nonce": "2c128e4d-fc91-4cd3-86b8-18bdea0988cb",
      "wallet_nonce": "qPmxiNFCR3QTm19POc8u",
      "state": "3be39b69-6ac1-41aa-921b-3e6c07ddcb03",
      "iss": "https://relying-party.example.org",
      "iat": 1672418465,
      "exp": 1672422065,
      "request_uri_method": "post"
    }

> **Steps 15-17 (WI Checks)**: The Request Object, in the form of signed JWT, is verified by the Wallet Instance. The Wallet Instance processes the Relying Party metadata and applies the policies related to the Relying Party, attesting whose Digital Credentials and User data the Relying Party is granted to request.

Reviewer 2: ```suggestion
**Steps 15-17 (WI Checks)**: The Wallet Instance verifies the Request Object, which is in the form of a signed JWT. It then processes the Relying Party metadata and applies the relevant policies to determine which Digital Credentials and User data the Relying Party is authorized to request.
```

**Steps 18-19 (User Consent)**: The Wallet Instance requests the User's consent to disclose the requested Credentials by showing the Relying Party's identity and the requested attributes. The User authorizes and consents the presentation of the Credentials by selecting/deselecting the personal data to release.

**Step 20 (Authorization Response)**: The Wallet Instance provides the Authorization Response to the Relying Party using an HTTP request with the method POST (response mode "direct_post.jwt").
  

>   Below is a non-normative example of the Authorization Response request:

Reviewer 2: ```suggestion
  Below is a non-normative example of the Authorization Response:
```

  .. code-block:: http

    POST /response_uri HTTP/1.1
    HOST: relying-party.example.org
    Content-Type: application/x-www-form-urlencoded
    
    response=eyJhbGciOiJFUzI1NiIs...9t2LQ
  
  Below is a non-normative example of the decrypted payload of the JWT contained in the ``response``, before base64url encoding. The ``vp_token`` parameter value corresponds to the format used when the DCQL query language is used in the presentation request.
  
  .. code-block:: json

    {
      "state": "3be39b69-6ac1-41aa-921b-3e6c07ddcb03",
      "vp_token": {
          "personal id data": "eyJhbGciOiJFUzI1NiIs...PT0iXX0",
          "wallet attestation": "eyJhbGciOiJFUzI1NiIs...NTi0XG"
      }
    }

> **Steps 21-25 (RP Checks)**: The Relying Party verifies the Authorization Response, extracts the Wallet Attestation to establish the trust with the Wallet Solution. The Relying Party extracts the Digital Credentials and attests the trust to the Credentials Issuer and the proof of possession of the Wallet Instance about the presented Digital Credentials. Finally, the Relying Party verifies the revocation status of the presented Digital Credentials as described in :ref:`Digital Credential Revocation and Suspension`. If all previous verifications yelded positive result, the Relying Party updates the User session.

Reviewer 2: ```suggestion
**Steps 21-25 (RP Checks)**: The Relying Party verifies the Authorization Response, extracts the Wallet Attestation to establish trust with the Wallet Solution. It then extracts the Digital Credentials and attests trust with the Credentials Issuer and the Wallet Instance's proof of possession of the presented Digital Credentials. Finally, the Relying Party verifies the revocation status of the presented Digital Credentials as described in :ref:`Digital Credential Revocation and Suspension`. If all previous verifications yelded positive result, the Relying Party updates the User session.
```

**Steps 26-27 or 28 (Relying Party Response)**: The Relying Party provides to the Wallet Instance the response about the presentation, which informs the User.
  
  In the **Same Device Flow**, the Relying Party SHOULD also pass the ``redirect_uri`` to the Wallet Instance. Upon receiving the ``redirect_uri``, the Wallet Instance MUST perform a redirect to the URL specified by the ``redirect_uri``. This redirect allows the Relying Party to seamlessly resume interaction with the User on the device which initiated the flow. When the response does not contain the ``redirect_uri`` parameter, the Wallet Instance is not required to perform any further step. The User should manually close the Wallet Instance and open the user-agent to continue the flow. 
  
  The following is a non-normative example of the response from the Relying Party to the Wallet Instance upon receiving the Authorization Response at the Response Endpoint.

  .. code-block:: http

    HTTP/1.1 200 OK
    Content-Type: application/json

    {
      "redirect_uri": "https://relying-party.example.org/cb?response_code=091535f699ea575c7937fa5f0f454aee"
    }

> 

Reviewer 2: for sake of clarity, I would also specify that in the cross device flow the response will always be a 200 ok without the redirect which is passed by the status endpoint.

> **Steps 29-30**: The JavaScript page inspecting the status endpoint.

Reviewer 2: ```suggestion
**Steps 29-30**: The JavaScript page is inspecting the status endpoint.
```

  
  Below is a non-normative example of the HTTP Request to the status endpoint, where the parameter ``id`` contains an opaque and random value:

  .. code-block:: http 

    GET /session-state?id=3be39b69-6ac1-41aa-921b-3e6c07ddcb03 HTTP/1.1
    HOST: relying-party.example.org
  
  When the Wallet Instance has provided the presentation to the Relying Party's  **response_uri** endpoint and the User authentication is successful. The Relying Party updates the session cookie allowing the user-agent to access to the protected resource. A redirect URL is provided carrying the location where the user-agent is intended to navigate.
  The following is a non-normative example of the response with the ``redirect_uri`` from the Relying Party to the user-agent.

  .. code-block:: http

    HTTP/1.1 200 OK
    Content-Type: application/json

    {
      "redirect_uri": "https://relying-party.example.org/cb?response_code=091535f699ea575c7937fa5f0f454aee"
    }

**Steps 31-32**: The user-agent is redirected to the redirect URI to continue the navigation with the protected resource made available to the User.

Authorization Request
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The URL parameters contained in the Relying Party Authorization Request, containing the request URI where to download the signed Request Object, are described in the Table below.

Reviewer 2: ```suggestion
The URL parameters contained in the Relying Party Authorization Request, which include the ``request_uri`` where the signed Request Object can be downloaded, are described in the table below.
```

The following is an example of an error response from ``request_uri`` endpoint:

.. code-block:: http

  HTTP/1.1 500 Internal Server Error

>   Content-Type: application/json

Reviewer 2: line 465 "The request_uri cannot be retrieved due to an internal server error." --> "The request object cannot be retrieved due to an internal server error."

    - Unique identifier provided by the Relying Party within the Authorization Request.

> SD-JWT defines how an Holder can present a Digital Credential to a Relying Party proving the legitimate possession of the Digital Credential. For doing this the Holder MUST include the ``KB-JWT`` in the SD-JWT, by appending the ``KB-JWT`` at the end of the of the SD-JWT, as represented in the example below:

Reviewer 2: ```suggestion
SD-JWT defines how a Holder can present a Digital Credential to a Relying Party, proving the legitimate possession of the Digital Credential. To do this, the Holder MUST include the ``KB-JWT`` in the SD-JWT by appending the ``KB-JWT`` at the end of the SD-JWT, as represented in the example below
```

    - REQUIRED. The base64url-encoded hash digest over the Issuer-signed JWT and the selected disclosures. 

Authorization Response Errors
-----------------------------

There are cases where the Wallet Instance cannot validate the Request Object or the Request Object results invalid. This error occurs if the Request Object is successfully fetched from the url provided in the parameter ``request_uri`` but fails the validation checks. This could be due to incorrect signatures, malformed claims, or other validation failures, such as the revocation of the Relying Party.

> If the Wallet Instance encounters any error during the evaluation of the Authorization Request, it MUST notify the Relying Party by sending an Authorization Error Response. 

Reviewer 2: ```suggestion
If the Wallet Instance encounters any such errors during the evaluation of the Authorization Request, it MUST notify the Relying Party by sending an Authorization Error Response. 
```

The Relying Party MUST include a response code within the ``redirect_uri``. The response code is a fresh, cryptographically random number used to ensure only the receiver of the redirect can fetch and process the Authorization Response. The number could be added as a path component, as a parameter or as a fragment to the URL. It is RECOMMENDED to use a cryptographic random value of 128 bits or more at the time of the writing of this specification.
Even if an adversary manages to steal the random value used in the request to the status endpoint, their user-agent would be rejected due to the missing cookie in the request.

.. warning::
    For security reasons and to prevent endpoint mix-up attacks, the value contained in the ``redirect_uri`` parameter MUST be one of those attested by a trusted third party, such as those provided in the ``openid_credential_verifier`` metadata within the ``redirect_uris`` parameter, obtained from the Trust Chain about the Relying Party.

Relying Party Response Errors
--------------------------------

> If any validation check performed by the Relying Party on the Authorization Response from the Wallet Instance fails, the Response URI endpoint MUST return an error response. The structure of this error response should be determined by the specific nature of the error encountered. The response MUST use ``application/json`` as the content type and MUST include the following parameters:

Reviewer 2: ```suggestion
If any validation check, performed by the Relying Party on the Authorization Response from the Wallet Instance, fails; the Response URI endpoint MUST return an error response. The structure of this error response should be determined by the specific nature of the error encountered. The response MUST use ``application/json`` as the content type and MUST include the following parameters:
```

    "error_description": "The vp_token is malformed, missing required parameters or incorrectly formatted"
  }

     

Status Endpoint
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> This specification introduces the Relying Party Status Endpoint for implementations that want to use it, as this endpoint is an internal feature for the security of the implementation and is not required for interoperability.

Reviewer 2: ```suggestion
This specification introduces the Relying Party Status Endpoint for implementations that choose to use it. This endpoint is an internal security feature of the implementation and is not required for interoperability.
```

> Be the flow Same Device or Cross Device, the user-agent needs to check the session status to the endpoint made available by Relying Party (status endpoint).

Reviewer 2: ```suggestion
Whether the flow is Same Device or Cross Device, the user-agent needs to check the session status at the endpoint made available by the Relying Party (status endpoint).
```

This check MAY be implemented in the form of JavaScript code, within the page that shows the QRCode or the href button pointing to the request URL.

> The JavaScript code makes the user-agent check the status endpoint using a polling strategy in seconds or a push strategy (e.g., web socket).

Reviewer 2: ```suggestion
The JavaScript code makes the user-agent check the status endpoint using either a polling strategy (in seconds) or a push strategy (e.g., WebSocket).
```

> Since the html page and the status endpoint are implemented by the Relying Party, it is under the Relying Party responsability the implementation details of this solution, because this is related to the Relying Party's internal API. However, the text below describes an implementation example.

Reviewer 2: ```suggestion
Since the HTML page and the status endpoint are implemented by the Relying Party, the implementation details of this solution are the responsibility of the Relying Party, as this is related to the Relying Party's internal API. However, the text below describes an example implementation.
```

Status Endpoint Errors
------------------------

If instead any validation check performed by the Relying Party fails, the QRCode page SHOULD be updated with an error message. Moreover, the status endpoint MUST return an error response, whose structure depends on the nature of the error. The response MUST use ``application/json`` as the content type and MUST include the following parameters:

* ``error``: The error code.
* ``error_description``: Text in human-readable form providing further details to clarify the nature of the error encountered.

The following table lists the HTTP Status Codes and related error codes that MUST be supported for the error response:

.. list-table:: 
    :widths: 20 20 60
    :header-rows: 1

    * - **Status Code**
      - **Error Code**
      - **Description**
    * - ``401 Unauthorized``
      - ``authentication_failed``
      - The Wallet Instance or its User have rejected the request, the request is expired, or other errors prevented the authentication. 
    * - ``403 Forbidden``
      - ``invalid_session``
      - Either the session id provided in the request is invalid.

> Redirect URI 

Reviewer 2: suggestion: move the Redirect URI section before the Status endpoint section, the latter is, in fact, a non normative section and does somehow break the flow description inserting a non-normative implementation description in between.
